src/Microsoft.DotNet.Build.CloudTestTasks/FilterForUpload.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.CloudTestTasks
+{
+    public sealed class FilterForUpload : Task
+    {
+        [Required]
+        public string TestOutputDirListsLocation { get; set; }
+
+
+        [Required]
+        public string FilterToTestTFM { get; set; }
+
+        [Required]
+        public string FilterToOSGroup { get; set; }
+
+        [Output]
+        public ITaskItem[] FilteredUploadItems
+        {
+            get { return _copyLocalItems.ToArray(); }
+        }
+
+        private readonly List<ITaskItem> _copyLocalItems = new List<ITaskItem>();
+
+        public override bool Execute()
+        {
+
+            Debugger.Launch();
+            string[] filesToSearch = Directory.GetFiles(TestOutputDirListsLocation);
+            
+            foreach (var filename in filesToSearch)
+            {
+                string selectedTest = string.Empty;
+                using (StreamReader sr = new StreamReader(File.OpenRead(filename)))
+                {
+                    string line;
+                    while ((line = sr.ReadLine()) != null)
+                    {
+                        if (calculateMatchValue(line) > calculateMatchValue(selectedTest))
+                        {
+                            int lastIndexOfSlash = line.LastIndexOf('/');
+                            string path = line.Substring(0, lastIndexOfSlash);
+                            string testTfm = line.Substring(lastIndexOfSlash + 1);
+                            lastIndexOfSlash = path.LastIndexOf('/');
+                            string testArchiveDir = Path.Combine(
+                                path.Substring(0, lastIndexOfSlash),
+                                "archive",
+                                "tests");
+                            string testZipName = path.Substring(lastIndexOfSlash + 1) + ".zip";
+                            selectedTest = Path.Combine(testArchiveDir, testTfm, testZipName);
+                        }
+                    }
+                }
+                if (!string.IsNullOrEmpty(selectedTest))
+                {
+                    _copyLocalItems.Add(new TaskItem(selectedTest));
+                }
+            }
+            return true;
+        }
+        
+        private int calculateMatchValue(string lines)
+        {
+            int matchValue = 0;
+            if (string.IsNullOrEmpty(lines))
+            {
+                return -1;
+            }
+            if (lines.Contains(FilterToOSGroup))
+            {
+                matchValue++;
+            }
+            if (lines.Contains(FilterToTestTFM))
```

Comment:
It was unexpected to me the first reading that not matching TestTFM returns -1, even though "matchValue" was incremented by FilterToOSGoup.

I think an early return of if(!lines.Contains(FilterToTestTFM)) return -1,
and then return lines.Contains(FilterToOSGroup) ? 2 : 1, makes it a bit easier to tell what this function returns in the various cases.